src/generator/qbe.rs

```diff
@@ -0,0 +1,1130 @@
+/**
+ * Copyright 2021 Alexey Yerin
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+use super::{Generator, GeneratorResult};
+use crate::ast::types::Type;
+use crate::ast::*;
+use std::collections::HashMap;
+
+pub struct QbeGenerator {
+    /// Counter for unique temporary names
+    tmp_counter: u32,
+    /// Block-scoped variable -> temporary mappings
+    scopes: Vec<HashMap<String, (QbeType, QbeValue)>>,
+    /// Structure -> (size, type, offset table) mappings
+    struct_map: HashMap<String, (u64, QbeType, OffsetTable)>,
+    /// Label prefix of the current loop
+    loop_label: Option<String>,
+    /// Data defintions collected during generation
+    datadefs: Vec<QbeDataDef>,
+}
+
+/// Mapping of field -> struct offset
+type OffsetTable = HashMap<String, u64>;
+
+impl Generator for QbeGenerator {
+    fn generate(prog: Module) -> GeneratorResult<String> {
+        let mut generator = QbeGenerator {
+            tmp_counter: 0,
+            scopes: Vec::new(),
+            struct_map: HashMap::new(),
+            loop_label: None,
```

Comment:
Does this work recursively?

Reply:
What exactly?

Reply:
Down [here](https://github.com/antimony-lang/antimony/pull/42/files#diff-d7fb403429bff994dfbfe09c033aaef341e51d88ab055a3a787e99a967783816R400) we clear the loop label when the while loop is done. I could imagine that this breaks nested loops. Does it?

```
// Obviously doesn't work, but you get the point
while true {
  while true {
    continue
  }
  continue
}
```

Reply:
Yeah, looks like it would break. I guess we can make a `Vec<String>` of labels and `push`/`pop` in loops

Reply:
Sounds reasonable